README.md
# ASC-Database

A program to keep track of one's session count with the ARRL VEC. The power of this program lies not in it's simple interface. Rather, the power resides in it's ability to generate reports.

## Current release version is 0.92.1

Read the ChangeLog for details.

Check the **Wiki** for more information.


ChangeLog
V 0.90.0  Initial program and upload

V 0.90.1  Some minor GUI fixes

V 0.91  Added Auto-Update functionality

V 0.91.1 Fixed lookup without entering a callsign

V 0.92  Added an Auto-Purge function to the Update function. This removes any VE who has lost 
        their ARRL accreditation from the database. Also reworked the Lookup button. If one 
        has registered their callsign, clicking on the button without entering a callsign 
        will automatically lookup one's statistics. Also reworked the Summary report. For
        1x2 or 2x1 callsigns, if there is a multiple match, one can add a '!' to the end
        of their callsign which will restrict the matching process. It's not perfect. Dropped
        the "Import by State" option which was for testing purposes.

V 0.92.1 Added a check for a new VE during an update

ASC-DB.py
#! /bin/python
## Bit and pieces are borrowed from my CVE-DB program
## Copyright 2023 Thomas Kocourek, N4FWD

import tkinter as tk
from tkinter import ttk
from tkinter import messagebox as mb

import os
import sys
import re
import sqlite3
from datetime import datetime
from croniter import croniter 
from threading import *
from threading import Thread
try:
    import pandas as pd
except:
    ## let's try an autoload here
    mb.showwarning("ERROR!","Python library 'pandas' is missing.\nRun command in terminal window:\npip install pandas")
    sys.exit()
    
## import support files here
import ve_utilities as ut
import global_var as gv
import asc_db_setup as adb
import read_session_count as rsc
import session_count_display as scd
import detailed_count_display as dcd
import set_up_defaults as sud

# for inter-thread comms
event = Event()

class Update(Thread):
    global event
    def __init__(self, app_self, func, display_widget, cron_data):
        super().__init__()
        self.keep_running = True
        self.cron_data = cron_data
        self.called_function = func
        self.display_widget = display_widget
        self.app_self = app_self
        self.no_skip = True
        
        self.display_widget.delete(1.0,tk.END)
        
    def stop(self):
        self.keep_running = False
        self.no_skip = False
    
            
    def run(self):
        curr_time = datetime.now()
        iter_time_obj = croniter(self.cron_data, curr_time)
        while self.keep_running:
            ## when 'iter_time_obj.get_next()' is invoked, the next available
            ## time is created. For a daily cron event, the date will jump
            ## to the next available day per the cron setting.
            curr_time = datetime.now()
            sometime = iter_time_obj.get_next(datetime)
            delta = (sometime - curr_time).total_seconds()
            text = "Auto-Update: Waiting time in seconds to next update: {}\n".format(delta)
            self.display_widget.insert(tk.END,text)
            event.wait(delta)
            if self.no_skip:
                text = "Auto-Update: Updating database at {}.\n".format(datetime.now())
                self.display_widget.insert(tk.END,text)
                self.called_function()
                text = "Auto-Update: Finished Update at {}.\n\n".format(datetime.now())
                self.display_widget.insert(tk.END,text)
            else:
                self.no_skip = True
            curr_time = datetime.now()
        text = "Auto-Update: Exiting thread.\n"
        self.display_widget.insert(tk.END,text)
              

class App(tk.Tk):
    def __init__(self):
        super().__init__()
        
        self.protocol("WM_DELETE_WINDOW",self.goodbye)
        self.environment = ut.set_environment
        ## GUI variables go next
        ## set up window dimensions, title , etc.
        self.title(gv.program)
        self.resizable(width=False, height=False)
        self.running = False
        self.update_db_obj = None
        ## save SQL results
        self.text_text = ""
        self.text_list = []
        #self.csv_dict_list = []
        
        self.as_of_var = tk.StringVar()
        self.ur_call_var = tk.StringVar()
        self.ur_county_var = tk.StringVar()
        self.ur_state_var = tk.StringVar()
        self.ur_session_count_var = tk.StringVar()
        self.ur_accreditation = tk.StringVar()
        
        self.their_call_var = tk.StringVar()
        self.their_county_var = tk.StringVar()
        self.their_state_var = tk.StringVar()
        self.their_session_count_var = tk.StringVar()
        self.their_accreditation = tk.StringVar()
        self.their_lookup_var = tk.StringVar()
        
        
        ## Let's start the window in the center of the desktop screen
        window_height = 700
        window_width = 1280
        screen_width = self.winfo_screenwidth()
        screen_height = self.winfo_screenheight()
        x_cordinate = int((screen_width/2) - (window_width/2))
        y_cordinate = int((screen_height/2) - (window_height/2))
        self.geometry("{}x{}+{}+{}".format(window_width, window_height, x_cordinate, y_cordinate))
        
        self.style = ttk.Style()
        self.style.theme_use("clam")
        
        
        self.columnconfigure(0, weight=1)
        self.columnconfigure(1, weight=1)
        self.columnconfigure(2, weight=1)
        #self.columnconfigure(3, weight=1)

        self.rowconfigure(0,weight=1)
        self.rowconfigure(1,weight=1)
        self.rowconfigure(2,weight=24)
        self.rowconfigure(3,weight=6)
        self.rowconfigure(4,weight=1)
        
        self.set_menu()
        self.set_up_ur_entries()
        self.set_gui()
        self.refresh_database()
        
    def set_menu(self):
        self.menubar = tk.Menu(self)
        self.filemenu = tk.Menu(self.menubar, tearoff = 0)
        self.filemenu.add_command(label="Import/Update All Session Counts",command=lambda: self.get_arrl_data())
        self.filemenu.add_command(label="Set Defaults", command=lambda: sud.set_defaults(self))
        self.filemenu.add_separator()
        self.filemenu.add_command(label="Exit", command=self.goodbye)
        
        ## menu items for reports
        ##
        self.reportmenu = tk.Menu(self.menubar, tearoff = 0)
        self.reportmenu.add_command(label="Your Session Count Summary", command=lambda: scd.session_count_data(self))
        self.reportmenu.add_command(label="Your Session Count Detailed", command=lambda: dcd.detailed_count_data(self))
        
        
        self.helpmenu = tk.Menu(self.menubar, tearoff = 0)
        self.helpmenu.add_command(label="About", command=self.about)
        
        self.menubar.add_cascade(label="File", menu = self.filemenu)
        self.menubar.add_cascade(label="Reports", menu = self.reportmenu)
        self.menubar.add_cascade(label="About", menu = self.helpmenu)
        self.config(menu = self.menubar)
        
        ## GUI widgets
    
    def set_gui(self):
        
        ## Labels for areas
        self.list_label = tk.Label(self, text="Results")
        self.list_label.grid(row=0, column=2, sticky='ne', padx=(10,350), pady=(10,10))
        
        self.result_text = tk.Text(self)
        self.result_text.grid(column=2, row=1, pady=(10,10), padx=(20,30), sticky='nes')
        self.result_text.configure(background="#d8f8d8", wrap="word", height=38, width=80,fg="#000000")
        self.result_text.delete(1.0,tk.END)
        self.text_scroll = ttk.Scrollbar(self, orient=tk.VERTICAL, command=self.result_text.yview)
        self.text_scroll.grid(column=2, row=1, sticky='nse', rowspan=20, pady=4, padx=(10,10))
        self.result_text['yscrollcommand'] = self.text_scroll.set
        
        self.urframe = tk.Frame(self)
        self.urframe.grid(row=0, column=0, rowspan=20, sticky='nsew', padx=10, pady=(0,10))
        self.urframe.columnconfigure(0,weight=1)
        self.urframe.columnconfigure(0,weight=5)
        
        self.as_of_label = tk.Label(self.urframe, text='Status as of:')
        self.as_of_label.grid(column=0, row=0, sticky='ne', padx=(5,5), pady=(5,5))
        self.as_of_entry = tk.Entry(self.urframe, textvariable = self.as_of_var, width=30, state='readonly')
        self.as_of_entry.grid(column=1, row=0, sticky='ne', padx=(5,5), pady=(5,5))
        
        self.ur_call_label = tk.Label(self.urframe, text = 'Your Callsign')
        self.ur_call_label.grid(column=0, row=1, sticky='ne', padx=(5,5), pady=(5,5))
        self.ur_call_entry = tk.Entry(self.urframe, textvariable = self.ur_call_var, width=30, state='readonly')
        self.ur_call_entry.grid(column=1, row=1, sticky='ne', padx=(5,5), pady=(5,5))
        
        self.spacer_label = tk.Label(self.urframe, text=" ")
        self.spacer_label.grid(column=0, row=6, sticky='nw', padx=(5,5), pady=(5,5))
        self.spacer2_label = tk.Label(self.urframe, text=" ")
        self.spacer2_label.grid(column=0, row=7, sticky='nw', padx=(5,5), pady=(5,5))
        
        self.their_lookup_button = tk.Button(self.urframe, text='Lookup Call', command=self.get_lookup_data)
        self.their_lookup_button.grid(column=0, row=8, sticky='ne', padx=(5,5), pady=(5,5))
        self.their_lookup_entry = tk.Entry(self.urframe, textvariable=self.their_lookup_var, width=30)
        self.their_lookup_entry.grid(column=1, row=8, sticky='nw', padx=(5,5), pady=(5,5))
        
        self.spacer3_label = tk.Label(self.urframe, text=" ")
        self.spacer3_label.grid(column=0, row=9, sticky='nw', padx=(5,5), pady=(5,5))
        self.spacer4_label = tk.Label(self.urframe, text=" ")
        self.spacer4_label.grid(column=0, row=10, sticky='nw', padx=(5,5), pady=(5,5))
        
        self.their_call_label = tk.Label(self.urframe, text='Callsign')
        self.their_call_label.grid(column=0, row=11, sticky='ne', padx=(5,5), pady=(5,5))
        self.their_call_entry = tk.Entry(self.urframe, textvariable = self.their_call_var, width=30)
        self.their_call_entry.grid(column=1, row=11, sticky='ne', padx=(5,5), pady=(5,5))
        
        self.their_county_label = tk.Label(self.urframe, text = 'County')
        self.their_county_label.grid(column=0, row=12, sticky='ne', padx=(5,5), pady=(5,5))
        self.their_county_entry = tk.Entry(self.urframe, textvariable = self.their_county_var, width=30)
        self.their_county_entry.grid(column=1, row=12, sticky='ne', padx=(5,5), pady=(5,5))
        
        self.their_state_label = tk.Label(self.urframe, text = 'State')
        self.their_state_label.grid(column=0, row=13, sticky='ne', padx=(5,5), pady=(5,5))
        self.their_state_entry = tk.Entry(self.urframe, textvariable = self.their_state_var, width=30)
        self.their_state_entry.grid(column=1, row=13, sticky='ne', padx=(5,5), pady=(5,5))
        
        self.their_session_count_label = tk.Label(self.urframe, text = "Count")
        self.their_session_count_label.grid(column=0, row=14, sticky='ne', padx=(5,5), pady=(5,5))
        self.their_session_count_entry = tk.Entry(self.urframe, textvariable = self.their_session_count_var, width=30)
        self.their_session_count_entry.grid(column=1, row=14, sticky='ne', padx=(5,5), pady=(5,5))
        
        self.their_accreditation_label = tk.Label(self.urframe, text = 'Accredited')
        self.their_accreditation_label.grid(column=0, row=15, sticky='ne', padx=(5,5), pady=(5,5))
        self.their_accreditation_entry = tk.Entry(self.urframe, textvariable = self.their_accreditation, width=30)
        self.their_accreditation_entry.grid(column=1, row=15, sticky='ne', padx=(5,5), pady=(5,5))
        
        
        
    def refresh_database(self):
        #print("self.running: ",self.running)
        db_connection = sqlite3.connect(gv.asc_database)
        db_cursor = db_connection.cursor()
        db_cursor.execute("SELECT * FROM settings")
        self.settings_check = db_cursor.fetchone()
        #print("setting: ",self.settings_check)
        if len(gv.settings) == 0:
            self.ur_call_var.set(self.settings_check[1])
            self.as_of_var.set(self.settings_check[2])
        else:
            self.ur_call_var.set(gv.settings[1])
            self.as_of_var.set(gv.settings[2])
        
        if self.running == False and self.settings_check[4] == '1':
            
            ## Inserting default auto-update values into the 'settings' table
            rec_cols = ', '.join(gv.settings_field_update_list)
            q_marks = ','.join(list('?'*len(gv.settings_field_update_list)))
            values = tuple(gv.settings_default_update_values)
            sql = "INSERT INTO settings ("+rec_cols+") VALUES ("+q_marks+")"
            sql_result = db_cursor.execute(sql,values)
            ## commit insert
            db_connection.commit()
        
        ## Do not invoke the auto-update until parameters are set
        elif self.settings_check[4] == '1' and self.running : 
            if self.update_db_obj == None: ## thread obj has not been initialized
                #print("creating obj.")
                self.update_db_obj = Update(self, self.run_auto_update,self.result_text,gv.cron_string) ## Pass the parameters
                self.update_db_obj.start()
                self.update_idletasks()
        else:
            ## do these following items only if the thread is running
            if self.update_db_obj != None: ## object was initialized
                self.update_db_obj.stop()
                event.set() ## force threading event to release
                self.update_db_obj = None
                self.running = False
            ## Inserting default values into the 'settings' table
            tmp_sql = ""
            for item in gv.settings_field_update_list:
                tmp_sql += item+'= ?, '
            tmp_sql = tmp_sql[:-2]
            values = tuple(gv.settings_default_update_values)
            sql = "UPDATE settings SET "+tmp_sql
            db_cursor.execute(sql,values)
            
            ## commit insert
            db_connection.commit()
            
        db_connection.close()   
            
    def run_auto_update(self):
        self.get_quiet_arrl_data()

    def get_quiet_arrl_data(self):
        for state in gv.states_list:
            website = gv.arrl_url + state
            rsc.get_count(self,state)
        self.update_idletasks()
        
    def get_arrl_data(self):
        self.result_text.delete(1.0,tk.END)
        for state in gv.states_list:
            website = gv.arrl_url + state
            text = "Importing from URL: "+website+'\n'
            self.result_text.insert(tk.END,text)
            self.result_text.yview(tk.END)
            self.update_idletasks()
            rsc.get_count(self,state)
        text = "Finished."
        self.result_text.insert(tk.END,text)
        self.update_idletasks()
            
    def get_state_data(self):
        db_connection = sqlite3.connect(gv.asc_database)
        db_cursor = db_connection.cursor()

        db_cursor.execute("SELECT * FROM settings")
        setting = db_cursor.fetchone()
        db_connection.close()
        self.result_text.delete(1.0,tk.END)
        state = setting[3]
        website = gv.arrl_url + state
        text = "Importing from URL: "+website+'\n'
        self.result_text.insert(tk.END,text)
        self.update_idletasks()
        rsc.get_count(self,state)
        text = "Finished."
        self.result_text.insert(tk.END,text)
        self.update_idletasks()
    
    def set_up_ur_entries(self):
        self.get_data()
        self.ur_call_var.set(gv.settings[1])
        self.as_of_var.set(gv.settings[2])
        self.ur_county_var.set(gv.ve_stat[1])
        self.ur_state_var.set(gv.ve_stat[4])
        self.ur_session_count_var.set(gv.ve_stat[3])
        self.ur_accreditation.set(gv.ve_stat[2])
        
        self.their_call_var.set('')
        self.their_county_var.set('')
        self.their_state_var.set('')
        self.their_session_count_var.set('')
        self.their_accreditation.set('')

    def get_lookup_data(self):
        tmp_list = []
        sql_text = "SELECT * FROM ve_count WHERE call LIKE ?"
        lookup_callsign = self.their_lookup_var.get()
        ## was button clicked without entering a callsign?
        if lookup_callsign == 'NOCALL':
            return
        elif len(lookup_callsign) == 0:
            lookup_callsign = self.ur_call_var.get()
            self.result_text.delete(1.0,tk.END)
        ## check for no wildcard match condition
        if lookup_callsign[-1] == '!':
            lup_callsign = lookup_callsign[:-1].upper()+'%'
        else:
            lup_callsign = '%'+lookup_callsign.upper()+'%'
        
        #self.result_text.delete(1.0,tk.END)
        db_connection = sqlite3.connect(gv.asc_database)
        db_cursor = db_connection.cursor()
        
        tmp_list.append(lup_callsign)
        try:
            db_cursor.execute(sql_text,tuple(tmp_list))
            record_check = db_cursor.fetchall()
        except:
            record_check = None
        db_connection.close()
        
        self.their_lookup_var.set('')
        if record_check != None and record_check != []:
            if len(record_check) == 1:
                record = record_check[0]
                self.their_call_var.set(record[1])
                self.their_county_var.set(record[2])
                self.their_state_var.set(record[5])
                self.their_session_count_var.set(record[4])
                self.their_accreditation.set(record[3])
            else:
                ## more than one match, place in Text widget
                ## first blank out any previous result
                self.their_call_var.set('')
                self.their_county_var.set('')
                self.their_state_var.set('')
                self.their_session_count_var.set('')
                self.their_accreditation.set('')
                self.result_text.delete(1.0,tk.END)
                ## bubble sort on session count
                record_check = self.sort_list_of_tuples(record_check)
                for r in record_check:
                    text_line = "Count:{}, Call:{}, County:{}, State:{}\n".format(r[4],r[1],r[2],r[5])
                    self.result_text.insert(tk.END,text_line)
        else:
            if lookup_callsign[-1] == '!':
                ls = lookup_callsign[:-1]
                lookup_callsign = ls
            text = "Callsign {} was not found!".format(lookup_callsign.upper())+'\n'
            self.result_text.insert(tk.END,text)
        self.update_idletasks()
    
    def get_data(self):
        tmp_list = []
        db_connection = sqlite3.connect(gv.asc_database)
        db_cursor = db_connection.cursor()

        db_cursor.execute("SELECT * FROM settings")
        gv.settings = db_cursor.fetchone()
        try:
            callsign = '%'+gv.settins[1]+'%'
            tmp_list.append(callsign)
            sql = "SELECT * FROM ve_count WHERE call LIKE ?"
            db_cursor.execute(sql,tuple(tmp_list))
            gv.ve_stat = db_cursor.fetchall()
            ## remove sql reference to record
            gv.ve_stat.pop(0)
        except:
            gv.ve_stat = ['NO-CALL','NO-COUNTY','NO-ACCREDITATION','NO-COUNT','NO-STATE']
        
        db_connection.close()
    
    def sort_list_of_tuples(self,lt):
        ## bubble sort a list of tuples with highest count at the top
        list_length = len(lt)
        ## subtle reference to a JK flipflop circuit
        for j in range(0, list_length):
            for k in range(0, list_length-j-1):
                if (int(lt[k][4]) < int(lt[k + 1][4])):
                    temp_tuple = lt[k]
                    lt[k] = lt[k + 1]
                    lt[k + 1] = temp_tuple
        return lt
        
    def goodbye(self):
        if self.update_db_obj != None:
            self.update_db_obj.stop()
            event.set() ## force threading event to release
        sys.exit()
        
    def about(self):
        info = gv.program+"\n\n"
        info += "Version: "+gv.version+'\n\n'
        info += "Written by Thomas Kocourek, N4FWD\n\n"
        info += "Copyright 2023"+'\n\n'
        info += "Released under the GNU 3 License \n\n"
        mb.showinfo('About',info)   

def main():
    ## set-up database. adb is imported at beginning of this file
    environment = adb.setup()
    if not environment:
        rw = tk.Tk()
        rw.overrideredirect(1)
        rw.withdraw()
        mb.showwarning('Error','Problem with environment or permissions!')
        rw.destroy()
        sys.exit()
    app = App()
    app.mainloop()
    
if __name__ == '__main__':
    main()

asc_db_setup.py
import os
import sqlite3
import sys
import ve_utilities as ut
import global_var as gv
import tkinter as tk
from tkinter import messagebox as mb

def setup():
    ## get the correct path for working directory
    ## get_environment() only sets up the path
    ## and returns it
    basic_dir = ut.set_environment()
    
    ## Change to correct directory
    os.chdir(basic_dir)
    ## set the flag
    db_result_flag = True
    
    ## set global path to database
    tmp_dir = os.path.join(basic_dir,gv.asc_dir)
    gv.asc_database = os.path.join(tmp_dir,"asc.db")
    
    if os.path.exists(gv.asc_database):
        ## do not erase an existing database
        ## reset the auto_update function
        db_connection = sqlite3.connect(gv.asc_database)
        db_cursor = db_connection.cursor()
        ## Inserting default auto_update values into the 'settings' table
        tmp_sql = ""
        for item in gv.settings_field_update_list:
            tmp_sql += item+'= ?, '
        tmp_sql = tmp_sql[:-2]
        values = tuple(gv.settings_default_update_values)
        sql = "UPDATE settings SET "+tmp_sql
        db_cursor.execute(sql,values)
        ## commit insert
        db_connection.commit()
    else:
        try:
            # else create a new database
            with open(gv.asc_database,mode="w"):pass

            ## build database ONLY if database did not exist
            
            db_connection = sqlite3.connect(gv.asc_database)
            db_cursor = db_connection.cursor()
            
            ## will be populated by data from website
            sql = """
                    CREATE TABLE IF NOT EXISTS ve_count (id INTEGER PRIMARY KEY AUTOINCREMENT,
                    call TEXT NOT NULL UNIQUE, county TEXT NOT NULL, accredit TEXT NOT NULL, scount TEXT NOT NULL, state TEXT NOT NULL, tag TEXT);
                """
            sql_result = db_cursor.execute(sql)
            
            ## 'date' is for recording the last date the database was updated
            sql = """
                    CREATE TABLE IF NOT EXISTS settings (id INTEGER PRIMARY KEY AUTOINCREMENT, yourcall TEXT ,
                    date TEXT , defaultState TEXT, autoflag TEXT , cronMin TEXT , cronHr TEXT , cronDom TEXT , cronMon TEXT , 
                    cronDow TEXT);
                """
            sql_result = db_cursor.execute(sql)
            
            ## commit both sql statements and create tables
            db_connection.commit() 
            
            ## Inserting default values into the 'settings' table
            rec_cols = ', '.join(gv.settings_field_list)
            q_marks = ','.join(list('?'*len(gv.settings_field_list)))
            values = tuple(gv.settings_default_values)
            sql = "INSERT INTO settings ("+rec_cols+") VALUES ("+q_marks+")"
            sql_result = db_cursor.execute(sql,values)
            ## commit insert
            db_connection.commit()
               
            db_cursor.execute("SELECT * FROM settings;")
            sql_records = db_cursor.fetchall()
            
            ## 've_count' table will be empty
            ## 'settings' table will have default values
            ## And close the new database
            #db_connection.close()
        except sqlite3.Error as er: ## in case of a SQL error, let's get more info
            print('SQLite error: %s' % (' '.join(er.args)))
            print("Exception class is: ", er.__class__)
            print('SQLite traceback: ')
            exc_type, exc_value, exc_tb = sys.exc_info()
            print(traceback.format_exception(exc_type, exc_value, exc_tb))
    ## And close the new database
    db_connection.close()    
    ## whether we were successful or not, return a flag
    return db_result_flag



ve_utilities.py
import os
import base64
import bz2
import platform
import shutil

import global_var as gv
import icons_array as ia 

asc_dir = ""
home_dir = os.path.expanduser('~')
basic_dir = ""

def set_environment():
    global asc_dir, home_dir, basic_dir
    ### Some code to accomodate the development environment
    ### leaving this in should not affect the production version
    gv.platform_os = platform.system()
    basic_dir = os.getcwd()
    project_dir = basic_dir[-8:]
    
    ## Is the program running in my development environment?
    ##
    ## Build correct path to working directory and save as a global
    ## That way the correct path can be referenced elsewhere
    ##
    ## The flag 'first_pass' is set to true in 'global_var.py' at start of program
    if gv.first_pass:
        if project_dir == "Projects":
            ## Build a path referencing the home dir
            asc_dir = os.path.join(basic_dir,gv.asc_dir)
            project3_dir = os.path.join(asc_dir, gv.report_dir)
            gv.base_rpt_dir = os.path.join(home_dir, project3_dir)
            
            ## while the program is running, set 'first_pass' to False
            gv.first_pass = False
            
        elif project_dir != "Projects":
            basic_dir = home_dir
            asc_dir = os.path.join(home_dir,gv.asc_dir)
            gv.base_rpt_dir = os.path.join(asc_dir, gv.report_dir)
            
            ## while the program is running, set 'first_pass' to False
            gv.first_pass = False
        else:
            pass
        
        ## Create the asc directory
        try:
            os.mkdir(asc_dir)
        except: ## directory already exists, do nothing
            pass
        
        try:
            os.mkdir(gv.base_rpt_dir)
        except: ## directory already exists, do nothing
            pass
    
        make_launcher()
        
    ## return that path
    return basic_dir

def make_launcher():
    global asc_dir, home_dir, basic_dir
    
     ## Path to Desktop
    desktop_dir = os.path.join(home_dir,"Desktop")
    
    sys_platform = gv.platform_os
    
    file_dict={'licon':"database.svg",'wicon':"database.ico",'wdesk':"ASC-DB.lnk"}
    
    keys = file_dict.keys()
    for key in keys:
        file_name = os.path.join(asc_dir,file_dict[key])
        string_data = ia.local_array[key]
        byte_data = bytes(string_data,'utf-8')
        decoded_data = base64.b64decode(byte_data)
        bin_data = bz2.decompress(decoded_data)
        with open(file_name,"wb") as f:
            f.write(bin_data)
    
    if sys_platform == "Windows":
        link_file = file_dict['wdesk']
        launcher_file = os.path.join(asc_dir,link_file)
        launcher_dest = os.path.join(desktop_dir,link_file)
        if not os.path.exists(launcher_dest):
            shutil.copy2(launcher_file,launcher_dest)
            
    elif sys_platform  == "Linux":
        ## let's create a desktop launcher
            launcher = "ASC-DB.desktop"
            desktop_launcher_path = os.path.join(desktop_dir,launcher)
            ## get the path towhere js8msg2 is executing from
            exec_path = os.path.join(home_dir,"bin/ASC-DB")
            icon_picture_path = os.path.join(asc_dir,"database.svg")
            ## updating launcher internals
            with open(desktop_launcher_path, "w") as fh:
                fh.write("[Desktop Entry]\n")
                fh.write("Version="+gv.version+"\n")
                fh.write("Type=Application\n")
                fh.write("Terminal=false\n")
                fh.write("Icon="+icon_picture_path+'\n')
                fh.write("Icon[en_US]="+icon_picture_path+'\n')
                fh.write("Name[en_US]=ASC-DB\n")
                fh.write("Exec="+exec_path+'\n')
                fh.write("Comment[en_US]="+gv.program+"\n")
                fh.write("Name=ASC-DB\n")
                fh.write("Comment="+gv.program+"\n")
            os.chmod(desktop_launcher_path,0o755)

global_var.py
## program vars
program = "ARRL Session Count for VEs"
version = "0.92.1"

top_window_x = 1280
top_window_y = 700
platform_os = ""

asc_dir = "ASC"
first_pass = True

report_dir = "REPORTS"
## path to the database file which is set in 'arrl_db_setup.py'
asc_database = ""
base_rpt_dir = ""

states_list = ['AK','AL','AR','AZ','CA','CO','CT','DC','DE','FL','GA','HI','IA','ID','IL','IN','KS','KY','LA','MA','MD','ME','MI','MN','MO','MS','MT','NC','ND','NE','NH','NJ','NM','NV','NY','OH','OK','OR','PA','PR','RI','SC','SD','TN','TX','UT','VA','VT','WA','WI','WV','WY']

arrl_url = "https://www.arrl.org/ve-session-counts?state="  ## needs State appended to complete the URL

settings_field_list = ['yourcall','date','defaultState','autoflag','cronMin','cronHr','cronDom','cronMon','cronDow']
settings_field_update_list = ['autoflag','cronMin','cronHr','cronDom','cronMon','cronDow']
settings_default_values = ['NOCALL','NotYet','GA','0','0','3','*','*','*']
settings_default_update_values = ['0','0','3','*','*','*']
settings = []
set_tmp = []
ve_stat = []

ve_field_list = ['call','county','accredit','scount','state','tag']
ve_input_list = ['call','county','accredit','scount']

def_sort_key = '0'

default_state = ""
default_auto = "Disabled"

month_selection_list = ['*','January','February','March','April','May','June','July','August','September','October','November','December']
month_selection_dict = {'January':'1','February':'2','March':'3','April':'4','May':'5','June':'6','July':'7','August':'8','September':'9','October':'10','November':'11','December':'12','*':'*'}
month_reverse_dict={'1':'January','2':'February','3':'March','4':'April','5':'May','6':'June','7':'July','8':'August','9':'September','10':'October','11':'November','12':'December','*':'*'}
minute_selection_list = ['0','5','10','15','20','25','30','35','40','45','50','55']
hour_selection_list = ['0','1','2','3','4','5','6','7','8','9','10','11','12','13','14','15','16','17','18','19','20','21','22','23']
## use input via 'Entry' widget for day of month
day_of_week_selection_list = ['*','Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday']
day_of_week_selection_dict ={'Sunday':'0','Monday':'1','Tuesday':'2','Wednesday':'3','Thursday':'4','Friday':'5','Saturday':'6','*':'*'}
dow_reverse_dict={'0':'Sunday','1':'Monday','2':'Tuesday','3':'Wednesday','4':'Thursday','5':'Friday','6':'Saturday','*':'*'}

day_of_month_selection_list = ['*','1','2','3','4','5','6','7','8','9','10','11','12','13','14','15','16','17','18','19','20','21','22','23','24','25','26','27','28']

select_report_list = ['State Higher','State Equal','State Lower','Total Higher','Total Equal','Total Lower']
select_report_list_default = 'State Higher'

cron_string = "0 3 * * *"

#event = None